packages/devkit/src/generators/project-name-directory-utils.ts

```diff
@@ -0,0 +1,282 @@
+import { prompt } from 'enquirer';
+import type { ProjectType } from 'nx/src/config/workspace-json-project-json';
+import type { Tree } from 'nx/src/generators/tree';
+import { requireNx } from '../../nx';
+import {
+  extractLayoutDirectory,
+  getWorkspaceLayout,
+} from '../utils/get-workspace-layout';
+import { names } from '../utils/names';
+
+const { joinPathFragments, readJson, readNxJson } = requireNx();
+
+export type ProjectNameDirectoryFormat = 'as-provided' | 'derived';
+export type ProjectGenerationOptions = {
+  name: string;
+  projectType: ProjectType;
+  directory?: string;
+  importPath?: string;
+  nameDirectoryFormat?: ProjectNameDirectoryFormat;
+  rootProject?: boolean;
+};
+
+export type ProjectNamesAndDirectories = {
+  /**
+   * Normalized full project name, including scope if name was provided with
+   * scope (e.g., `@scope/name`, only available when `nameDirectoryFormat` is
+   * `as-provided`).
+   */
+  projectName: string;
+  /**
+   * Normalized project directory, including the layout directory if configured.
+   */
+  projectDirectory: string;
+  names: {
+    /**
+     * Normalized project name without scope. It's meant to be used when
+     * generating file names that contain the project name.
+     */
+    projectFileName: string;
+    /**
+     * Normalized project name without scope or directory. It's meant to be used
+     * when generating shorter file names that contain the project name.
+     */
+    projectSimpleName: string;
+  };
+  /**
+   * Normalized import path for the project.
+   */
+  importPath?: string;
+};
+
+type ProjectNameDirectoryFormats = {
+  'as-provided': {
+    description: string;
+    options: ProjectNamesAndDirectories;
+  };
+  derived?: {
+    description: string;
+    options: ProjectNamesAndDirectories;
+  };
+};
+
+export async function determineProjectNamesAndDirectories(
```

Comment:
nitpick: Is it confusing if change this to `determineProjectOptions`? I also think `determineProjectNameAndRoot` would be an OK name even though it returns `importPath` and stuff..